_pytest/cacheprovider.py

```diff
@@ -168,6 +170,41 @@ def pytest_sessionfinish(self, session):
             config.cache.set("cache/lastfailed", self.lastfailed)
 
 
+class NFPlugin(object):
+    """ Plugin which implements the --nf (run new-first) option """
+
+    def __init__(self, config):
+        self.config = config
+        self.active = config.option.newfirst
+        self.all_items = config.cache.get("cache/allitems", {})
+
+    def pytest_collection_modifyitems(self, session, config, items):
+        if self.active:
+            new_items = []
+            other_items = []
+            for item in items:
+                mod_timestamp = os.path.getmtime(str(item.fspath))
+                if self.all_items and item.nodeid not in self.all_items:
```

Comment:
Unfortunately `item.nodeid not in self.all_items` will be a linear search, which can be expensive if we have a large collection.

I think you should change `new_items` and `other_items` to a `dict(nodeid -> timestamp)` instead, given that you will be sorting them later anyway.

Also you can check just with `if item.nodeid not in self.all_items:` because and empty list or dict will return `False` for the containment check.

Reply:
the code is incorrect about the typo of all_items, it changes multiple times

Reply:
@RonnyPfannschmidt You mean I need to rename it to smth like "cached_items"?

Reply:
@nicoddemus 
```
I think you should change new_items and other_items to a dict(nodeid -> timestamp) instead, given that you will be sorting them later anyway.
```
In this case I will not know about new tests, isn't it? Only about modified files

Reply:
> In this case I will not know about new tests, isn't it? Only about modified files

I did not explain correctly sorry, I meant this:

```python
            new_items = {}
            other_items = {}
            for item in items:
                if item.nodeid not in self.all_items:
                    new_items[item.nodeid] = item
                else:
                    other_items[item.nodeid] = item
```

And you can get the timestamp during your sorting function:

```python
            items[:] = self._get_increasing_order(six.itervalues(new_items)) + \
                self._get_increasing_order(six.itervalues(other_items.values))
            ...

    def _get_increasing_order(self, items):
        return sorted(items, key=lambda item: item.fspath.mtime(), reverse=True)
```

Reply:
I have failed tests because dict is not sorted in python. It need to be a list

Reply:
@nicoddemus this is not working in py27. Every time different order:
```
test_2/test_2.py::test_1[1] PASSED
test_2/test_2.py::test_1[2] PASSED
test_1/test_1.py::test_1[2] PASSED
test_1/test_1.py::test_1[1] PASSED
```
I think it's better to save my code based on lists. What do you think?

Reply:
Try changing `new_items` and `other_items` to a `OrderedDict` instead, that should be equivalent to the previous code.